In [9]:
from simpledbf import Dbf5
import pandas as pd
import recordlinkage


# SETUP - Identificação dos arquivos


In [10]:
import os

# Melhorar filtragem de dados

def find_files(path):
# r=root, d=directories, f = files
    files = []
    for r, d, f in os.walk(path):
        for file in f:
            if '.dbf' in file:
                files.append(os.path.join(r, file))
    return files

path_sinasc_2017 = '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/'
path_sinasc_2016 = '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/'
path_sim = '../DATASUS_DADOS_1996_A_2017_DBF/SIM/2017'

sinasc_2017 = find_files(path_sinasc_2017)
sinasc_2016 = find_files(path_sinasc_2016)
sim_files = find_files(path_sim)

sinasc_files = sinasc_2016 + sinasc_2017
all_files = sinasc_files + sim_files

for f in all_files:
    print(f)

../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAC2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAL2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAM2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAP2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNBA2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNCE2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNDF2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNES2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNGO2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNMA2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNMG2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNMS2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNMT2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNPA2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNPB2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNPE2016.dbf
../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNPI2016.dbf
../DATASUS_DAD

In [11]:
dict_sinasc_files = {}
for files in sinasc_files:
    if files[-10:-8] in dict_sinasc_files:
        dict_sinasc_files[files[-10:-8]].append(files)
    else:
        dict_sinasc_files[files[-10:-8]] = [files]

print(dict_sinasc_files)

{'AC': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAC2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAC2017.dbf'], 'AL': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAL2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAL2017.dbf'], 'AM': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAM2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAM2017.dbf'], 'AP': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAP2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAP2017.dbf'], 'BA': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNBA2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNBA2017.dbf'], 'CE': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNCE2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNCE2017.dbf'], 'DF': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNDF2016.dbf', '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNDF2017.dbf'], 'ES': ['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNES2016.dbf', '../DATASUS_DADOS_199

In [6]:
excel = pd.read_excel('../CADMUN.xls')
excel.UFCOD

0        0
1       11
2       11
3       11
4       11
        ..
5647    52
5648    52
5649    52
5650    52
5651    53
Name: UFCOD, Length: 5652, dtype: int64

In [5]:
is_amapa = excel['UFCOD'] == 16
is_amapa

0       False
1       False
2       False
3       False
4       False
        ...  
5647    False
5648    False
5649    False
5650    False
5651    False
Name: UFCOD, Length: 5652, dtype: bool

In [ ]:
excel['Código Município Completo'] == 

In [6]:
amapa_mun = excel[is_amapa]
amapa_mun

MUNCOD  MUNCODDV SITUACAO  MUNSINP  MUNSIAFI                  MUNNOME  \
300  160000   1600000    IGNOR  25000.0       NaN  Município ignorado - AP   
301  160005   1600055    ATIVO  25010.0     665.0           Serra do Navio   
302  160010   1600105    ATIVO  25001.0     601.0                    Amapá   
303  160015   1600154    ATIVO  25011.0     663.0  Pedra Branca do Amapari   
304  160020   1600204    ATIVO  25002.0     603.0                 Calçoene   
305  160021   1600212    ATIVO  25012.0     667.0                   Cutias   
306  160023   1600238    ATIVO  25006.0     611.0           Ferreira Gomes   
307  160025   1600253    ATIVO  25013.0     669.0                  Itaubal   
308  160027   1600279    ATIVO  25007.0     613.0         Laranjal do Jari   
309  160030   1600303    ATIVO  25003.0     605.0                   Macapá   
310  160040   1600402    ATIVO  25004.0     607.0                  Mazagão   
311  160050   1600501    ATIVO  25005.0     609.0                 Oiapoque   
312  160053   1600535    ATIVO  25014.0     671.0             Porto Grande   
313  160055   1600550    ATIVO  25015.0     673.0                 Pracuúba   
314  160060   1600600    ATIVO  25008.0     615.0                  Santana   
315  160070   1600709    ATIVO  25009.0     617.0          Tartarugalzinho   
316  160080   1600808    ATIVO  25016.0      70.0          Vitória do Jari   

                    MUNNOMEX                     OBSERV MUNSINON MUNSINONDV  \
300  MUNICIPIO IGNORADO - AP                        NaN  ,169999   ,1699999   
301           SERRA DO NAVIO  Ex Água Branca do Amapari      NaN        NaN   
302                    AMAPA                        NaN      NaN        NaN   
303  PEDRA BRANCA DO AMAPARI                 Ex Amapari      NaN        NaN   
304                 CALCOENE                        NaN      NaN        NaN   
305                   CUTIAS                        NaN      NaN        NaN   
306           FERREIRA GOMES                        NaN      NaN        NaN   
307                  ITAUBAL                        NaN      NaN        NaN   
308         LARANJAL DO JARI                        NaN      NaN        NaN   
309                   MACAPA                        NaN      NaN        NaN   
310                  MAZAGAO                        NaN      NaN        NaN   
311                 OIAPOQUE                        NaN      NaN        NaN   
312             PORTO GRANDE                        NaN      NaN        NaN   
313                 PRACUUBA                        NaN      NaN        NaN   
314                  SANTANA                        NaN      NaN        NaN   
315          TARTARUGALZINHO                        NaN      NaN        NaN   
316          VITORIA DO JARI                        NaN      NaN        NaN   

     ... CSAUDCOD RMETRCOD AGLCOD  ANOINST  ANOEXT  SUCESSOR  LATITUDE  \
300  ...    16000     1600   1600      NaN     NaN       NaN     0.000   
301  ...    16900     1690   1690   1993.0     NaN       NaN     0.896   
302  ...    16900     1690   1690   1956.0     NaN       NaN     2.053   
303  ...    16900     1690   1690   1993.0     NaN       NaN     0.778   
304  ...    16900     1690   1690   1956.0     NaN       NaN     2.498   
305  ...    16900     1690   1690   1993.0     NaN       NaN     0.986   
306  ...    16900     1690   1690   1989.0     NaN       NaN     0.858   
307  ...    16900     1690   1690   1993.0     NaN       NaN     0.596   
308  ...    16900     1690   1690   1989.0     NaN       NaN    -0.842   
309  ...    16900     1690   1601   1943.0     NaN       NaN     0.039   
310  ...    16900     1690   1690   1841.0     NaN       NaN    -0.115   
311  ...    16900     1690   1690   1945.0     NaN       NaN     3.843   
312  ...    16900     1690   1690   1993.0     NaN       NaN     0.713   
313  ...    16900     1690   1690   1993.0     NaN       NaN     1.743   
314  ...    16900     1690   1690   1989.0     NaN       NaN    -0.

In [70]:
try_base = excel['Código Município Completo'] == 1600105
try_base.unique()

array([False,  True])

In [7]:
cod_mun_amapa

NameError: name 'cod_mun_amapa' is not defined

In [5]:
dict_sinasc_files['AP']

['../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2016/DNAP2016.dbf',
 '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAP2017.dbf']

# Exploração - SINASC

São explorados as bases contendo dados para o estado do Amapá. É feito inicialmente uma exploração somente com o ano de 2017 para inicialmente entender melhor a base antes de unificar o restante da base.

dict de paths: dict_sinasc_files

ex:

```python
: dict_sinasc_files['AP']
> '../DATASUS_DADOS_1996_A_2017_DBF/SINASC/2017/DNAP2017.dbf'
```

In [24]:
# Need to analyze know how many fields are the same
dbf_sinasc_2016 = Dbf5(dict_sinasc_files['AP'][0], codec='utf-8')
dbf_sinasc_2017 = Dbf5(dict_sinasc_files['AP'][1], codec='utf-8')

df_list = []

for file in dict_sinasc_files['AP']:
    dbf_sinasc = Dbf5(file, codec='utf-8')
    df_sinasc_temp = dbf_sinasc.to_dataframe()
    df_list.append(df_sinasc_temp)
    
df_sinasc = pd.concat(df_list)


dbf_sim_1 = Dbf5(sim_files[1], codec='utf-8')
df_sim = dbf_sim_1.to_dataframe()

## Filtrando a base sim por rows pertecentes aos municipios avaliados

In [29]:
df_sinasc.CODMUNRES.unique()
cod_mun_res = [
    '160080', '160027', '160030', '160060',
    '160005', '160020', '160010', '160055',
    '160040', '160015', '160053', '160021',
    '160023', '160070', '160025', '160050'
]
df_sim_filtered = df_sim['CODMUNRES'].isin(cod_mun_res)
sim_ap = df_sim[df_sim_filtered]    

In [51]:
print('Tentativa de filtro de link por numerodo e numerodn nas bases sim e sinasc')
print('---------------------------------sim--------------------------------------')
print('--------------------------------------------------------------------------')
print('Ano 2017')
print('SIM com : ', len(sim_ap), ' rows.')
print('Quantidade de chaves com sinasc faltantes: ', sim_ap['NUMERODN'].isna().sum())
print('NUMERODO')
print('--------------------------------------------------------------------------')
print('-------------------------------sinasc-------------------------------------')
print('Ano 2016-2017')
print('SINASC : ', len(df_sinasc), ' rows.')
print('NUMERODN')


Tentativa de filtro de link por numerodo e numerodn nas bases sim e sinasc
---------------------------------sim--------------------------------------
--------------------------------------------------------------------------
Ano 2017
SIM com :  302  rows.
Quantidade de chaves com sinasc faltantes:  85
NUMERODO
--------------------------------------------------------------------------
-------------------------------sinasc-------------------------------------
Ano 2016-2017
SINASC :  30920  rows.
NUMERODN


In [53]:
print('Chaves estrangeiras faltosas sim: ', sim_ap['NUMERODN'].isna().sum())
print('Chaves estrangeiras faltosas sinasc: ', df_sinasc['NUMERODN'].isna().sum())

Chaves estrangeiras faltosas sim:  85
Chaves estrangeiras faltosas sinasc:  0


In [76]:
sinasc_key_filtered = df_sinasc['NUMERODN'].isin(list(sim_ap['NUMERODN']))
dead_sinasc = df_sinasc[sinasc_key_filtered]
new_sinasc = df_sinasc[~sinasc_key_filtered]
print(len(sim_ap))
print(len(df_sinasc))
print('--------------')
print('dead_sinasc_match: ', len(dead_sinasc))
print('alive_sinasc: ', len(new_sinasc))

302
30920
--------------
dead_sinasc_match:  200
alive_sinasc:  30720


### Considerações

dos 302 rows da base `SIM`, 217 tinha o NUMERODN cadastrado, que é a Declaração de nascido vivo, podendo ser usada para conectar as duas bases.

Dos 217 com NUMERODN, 200 foram corretamente conectados, os 17 que restaram, acredita-se não serem cadastrados na base `SINASC` do Amapá.

### Rows com NUMERODN faltosos

85 rows não possuem o NUMERODN cadastrado.
Para esses rows será utilizado o linkage na tentativa de encontrar mais conexões.

# LINKAGE

In [77]:
# Filtrando os rows que devem ser avaliados.
missing_sim = sim_ap[sim_ap['NUMERODN'].isna()]
missing_sim["PESO"] = pd.to_numeric(missing_sim["PESO"])
new_sinasc["PESO"] = pd.to_numeric(new_sinasc["PESO"])

new_sinasc.columns

/usr/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


Index(['NUMERODN', 'CODINST', 'ORIGEM', 'NUMERODV', 'PREFIXODN', 'CODESTAB',
       'CODMUNNASC', 'LOCNASC', 'IDADEMAE', 'ESTCIVMAE', 'ESCMAE',
       'CODOCUPMAE', 'QTDFILVIVO', 'QTDFILMORT', 'CODMUNRES', 'GESTACAO',
       'GRAVIDEZ', 'PARTO', 'CONSULTAS', 'DTNASC', 'HORANASC', 'SEXO',
       'APGAR1', 'APGAR5', 'RACACOR', 'PESO', 'IDANOMAL', 'DTCADASTRO',
       'CODANOMAL', 'NUMEROLOTE', 'VERSAOSIST', 'DTRECEBIM', 'DIFDATA',
       'DTRECORIGA', 'NATURALMAE', 'CODMUNNATU', 'CODUFNATU', 'ESCMAE2010',
       'SERIESCMAE', 'DTNASCMAE', 'RACACORMAE', 'QTDGESTANT', 'QTDPARTNOR',
       'QTDPARTCES', 'IDADEPAI', 'DTULTMENST', 'SEMAGESTAC', 'TPMETESTIM',
       'CONSPRENAT', 'MESPRENAT', 'TPAPRESENT', 'STTRABPART', 'STCESPARTO',
       'TPNASCASSI', 'TPFUNCRESP', 'TPDOCRESP', 'DTDECLARAC', 'ESCMAEAGR1',
       'STDNEPIDEM', 'STDNNOVA', 'CODPAISRES', 'TPROBSON', 'PARIDADE',
       'KOTELCHUCK'],
      dtype='object')

In [162]:
from recordlinkage.index import Block
from recordlinkage import Compare

indexer = Block(
    left_on=['CODMUNNATU', 'DTNASC'],
    right_on=['CODMUNNATU', 'DTNASC']
)

"""
Replicando: http://www.scielo.br/pdf/rbsmi/v15n1/1519-3829-rbsmi-15-01-0057.pdf
sexo da criança: OK
idade da mãe: OK 
raça/cor: OK 
escolaridade materna: OK 
ocupação da mãe: OK
número de filhos: OK
nascidos vivos: OK
número de filhos nascidos mortos: Não
tipo de parto: Ok
duração da gestação: Ok 
peso ao nascer: Ok
tipo de gravidez: Ok
"""

#new_sinasc = new_sinasc.reset_index()
candidate_links = indexer.index(missing_sim, new_sinasc)

print(len(candidate_links))
compare = Compare()


compare.exact('PARTO', 'PARTO', label='PARTO')
compare.exact('RACACOR', 'RACACOR', label='RACACOR')
compare.exact('SEXO', 'SEXO', label='SEXO')
compare.exact('IDADEMAE', 'IDADEMAE', label='IDADEMAE')
compare.numeric('PESO', 'PESO', label='PESO')
compare.exact('SEMAGESTAC', 'SEMAGESTAC', label='SEMAGESTAC')
compare.exact('QTDFILMORT', 'QTDFILMORT', label='QTDFILMORT')
compare.exact('QTDFILVIVO', 'QTDFILVIVO', label='QTDFILVIVO')
compare.exact('ESCMAE', 'ESCMAE', label='ESCMAE')
#compare.exact('OCUPMAE', 'OCUPMAE', label='OCUPMAE')
compare.exact('GRAVIDEZ', 'GRAVIDEZ', label='GRAVIDEZ')

#compare.exact('NUMERODO', 'NUMERODN', label='NUMERO')

compare_vectors = compare.compute(candidate_links, missing_sim, new_sinasc)

compare_vectors.describe()

1018


PARTO      RACACOR         SEXO     IDADEMAE         PESO  \
count  1018.000000  1018.000000  1018.000000  1018.000000  1018.000000   
mean      0.327112     0.569745     0.500000     0.039293     0.017682   
std       0.469389     0.495355     0.500246     0.194386     0.131857   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     1.000000     0.500000     0.000000     0.000000   
75%       1.000000     1.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

        SEMAGESTAC   QTDFILMORT   QTDFILVIVO       ESCMAE     GRAVIDEZ  
count  1018.000000  1018.000000  1018.000000  1018.000000  1018.000000  
mean      0.053045     0.341847     0.096267     0.194499     0.570727  
std       0.224234     0.474562     0.295102     0.396009     0.495216  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     1.000000  
75%       0.000000     1.000000     0.000000     0.000000     1.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000

In [163]:
len(compare_vectors)

1018

In [164]:
compare_vectors.sum(axis=1).value_counts().sort_index(ascending=False)

9.0      1
8.0      7
7.0     13
6.0     41
5.0     97
4.0    182
3.0    178
2.0    205
1.0    200
0.0     94
dtype: int64

In [166]:
matches = compare_vectors[compare_vectors.sum(axis=1) > 6]
print(len(matches))
matches.head(10)

21


PARTO  RACACOR  SEXO  IDADEMAE  PESO  SEMAGESTAC  QTDFILMORT  \
1057  18226      1        1     1         0   0.0           0           1   
2196  11849      1        1     1         0   0.0           1           1   
      11878      1        1     1         0   1.0           1           1   
14611 17022      1        1     1         1   1.0           0           0   
5610  28413      1        0     1         1   1.0           0           1   
6556  16485      1        0     1         0   0.0           1           1   
7319  11748      1        1     1         1   1.0           1           1   
      11818      1        1     1         0   0.0           1           1   
      11822      1        1     0         0   0.0           1           1   
7320  15325      1        1     1         0   0.0           1           1   

             QTDFILVIVO  ESCMAE  GRAVIDEZ  
1057  18226           1       1         1  
2196  11849           0       1         1  
      11878           0       1         1  
14611 17022           0       1         1  
5610  28413           1       0         1  
6556  16485           1       1         1  
7319  11748           0       1         1  
      11818           0       1         1  
      11822           1       1         1  
7320  15325           0       1         1

In [179]:
class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

In [180]:
same_fields = df_sim.columns.isin(df_sinasc.columns)
match_fields = list(df_sim.columns[same_fields])


def compare_matches(sim_index, sinasc_index):
    for field in match_fields:
        if missing_sim.loc[sim_index][field] == new_sinasc.loc[sinasc_index][field]:
            print(bcolors.OKGREEN, "field ", field, ': ', missing_sim.loc[sim_index][field], new_sinasc.loc[sinasc_index][field], bcolors.ENDC)
        else:
            print("field ", field, ': ', missing_sim.loc[sim_index][field], new_sinasc.loc[sinasc_index][field])

compare_matches(9029, 19753)
print('---------validar:----------')
compare_matches(2196, 11849)
print('-------------------')
compare_matches(2196, 11878)
print('-------------------')


field  CODINST :  MAP1600300001 MAP1600300003
 field  NUMERODV :  4 4 
 field  ORIGEM :  1 1 
 field  CODMUNNATU :  160030 160030 
 field  DTNASC :  29062017 29062017 
 field  SEXO :  1 1 
 field  RACACOR :  4 4 
 field  CODMUNRES :  160030 160030 
 field  CODESTAB :  2020068 2020068 
 field  IDADEMAE :  19 19 
 field  ESCMAE :  4 4 
 field  ESCMAE2010 :  3 3 
field  SERIESCMAE :  nan 3
field  QTDFILVIVO :  01 00
 field  QTDFILMORT :  00 00 
 field  GRAVIDEZ :  1 1 
field  SEMAGESTAC :  24 25
 field  GESTACAO :  2 2 
 field  PARTO :  1 1 
 field  PESO :  720.0 720.0 
field  NUMERODN :  nan 71431293
field  NUMEROLOTE :  20170036 20170018
field  DTCADASTRO :  03102017 30062017
field  VERSAOSIST :  3.2.00 3.2.01
field  DTRECEBIM :  16102017 03072017
field  DTRECORIGA :  16102017 nan
field  ESCMAEAGR1 :  12 06
field  DIFDATA :  036 004
---------validar:----------
field  CODINST :  MAP1600300007 MAP1600300003
field  NUMERODV :  8 3
 field  ORIGEM :  1 1 
 field  CODMUNNATU :  160030 160030 